### PR TITLE
Apply MVVM for global settings in SUI

### PR DESCRIPTION
**Pull request by carlos-zamora**
## Summary of the Pull Request
Introduces and uses `GlobalSettingsViewModel` as the view model for global settings in the settings UI. This makes it so that we don't actually store the `XXXNavigationState` on the page objects. Instead, we store the view model. The view model also stores some additional useful information like enum entry lists, current enum entry values, and getters for the settings model.

## References
#9207 - Apply MVVM

## Detailed Description of the Pull Request / Additional comments
Really this just involved me moving stuff into `GlobalSettingsViewModel`. There should be no change in functionality.

## Validation Steps Performed
Changed some global settings and saved/discarded changes. Verified correct behavior from that.

### CODE REVIEW

File: src/cascadia/TerminalSettingsEditor/GlobalAppearance.idl [hunk 1]
```diff
@@ -1,31 +1,19 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
 
+import "GlobalSettingsViewModel.idl";
 import "EnumEntry.idl";
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/GlobalAppearance.idl:5]: I can remove EnumEntry here

File: src/cascadia/TerminalSettingsEditor/GlobalSettingsViewModel.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+#include "GlobalSettingsViewModel.g.h"
+#include "Utils.h"
+#include "ViewModelHelpers.h"
+#include "..\TerminalSettingsModel\MTSMSettings.h"
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/GlobalSettingsViewModel.h:9]: I know... I really shouldn't just grab this from `TerminalSettingsModel`, but it lets me use the X-Macro and it's so nice! Should I move this to `inc` or something? I feel like it's fine.

File: src/cascadia/TerminalSettingsEditor/GlobalSettingsViewModel.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+#include "GlobalSettingsViewModel.g.h"
+#include "Utils.h"
+#include "ViewModelHelpers.h"
+#include "..\TerminalSettingsModel\MTSMSettings.h"
+#include "Utils.h"
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/GlobalSettingsViewModel.h:10]: whoops. I have Utils here twice haha.

File: src/cascadia/TerminalSettingsEditor/GlobalSettingsViewModel.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    public:
+        GlobalSettingsViewModel(const Model::GlobalAppSettings& globalSettings, const Model::CascadiaSettings& appSettings);
+
+        void SetInvertedDisableAnimationsValue(bool invertedDisableAnimationsValue);
+
+        _GET_SETTING_FUNC(_globals, DefaultProfile);
+        _SET_SETTING_FUNC(_globals, DefaultProfile);
+        _CLEAR_SETTING_FUNC(_globals, Language);
+
+        // Launch
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/GlobalSettingsViewModel.h:25]: (note to self) don't organize these by page. Organize them like this:
- custom getters/setters
- GETSET_BINDABLE_ENUM_SETTING
- X-Macro

File: src/cascadia/TerminalSettingsEditor/GlobalSettingsViewModel.idl [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+import "EnumEntry.idl";
+
+#include "ViewModelHelpers.idl.h"
+
+namespace Microsoft.Terminal.Settings.Editor
+{
+    runtimeclass GlobalSettingsViewModel : Windows.UI.Xaml.Data.INotifyPropertyChanged
+    {
+        // REGION: Launch
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/GlobalSettingsViewModel.idl:12]: I should get rid of this `// REGION` stuff and organize it like this:
- view model state properties & misc
- pairings of `CurrentXXX` and `XXXList`
- projected settings

File: src/cascadia/TerminalSettingsEditor/Launch.xaml [hunk 6]
```diff
@@ -74,8 +74,8 @@
                                         x:Uid="Globals_DefaultTerminal"
                                         x:Load="false">
                     <ComboBox x:Name="DefaultTerminal"
-                              ItemsSource="{x:Bind State.Settings.DefaultTerminals}"
-                              SelectedItem="{x:Bind State.Settings.CurrentDefaultTerminal, Mode=TwoWay}"
+                              ItemsSource="{x:Bind AppSettings.DefaultTerminals}"
+                              SelectedItem="{x:Bind AppSettings.CurrentDefaultTerminal, Mode=TwoWay}"
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/Launch.xaml:78]: This one is weird. We need `CascadiaSettings` for defterm. But `Globals` already has a reference to `CascadiaSettings`. For now, I've been passing it in separately to the navigation state and storing a reference to it on the page. Long term, we should be using `CascadiaSettingsViewModel` instead, I think. But idk if this abstraction is right. Thoughts?

File: src/cascadia/TerminalSettingsEditor/Microsoft.Terminal.Settings.Editor.vcxproj.filters [hunk 7]
```diff
@@ -22,6 +22,7 @@
       <Filter>Converters</Filter>
     </Midl>
     <Midl Include="SettingContainer.idl" />
+    <Midl Include="GlobalSettingsViewModel.idl" />
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/Microsoft.Terminal.Settings.Editor.vcxproj.filters:25]: (This would be a separate PR) Would there be any value to introducing these filters:
- view models
- views

And pulling out ProfileViewModel, KeyBindingViewModel, etc. into their own files and throwing them in the filter? Those classes are big enough that they could be their own objects tbh.

File: src/cascadia/TerminalSettingsEditor/Utils.h [hunk 8]
```diff
@@ -51,29 +51,29 @@
...
-    {                                                                                                                                                    \
-        return _##name##List;                                                                                                                            \
-    }                                                                                                                                                    \
-                                                                                                                                                         \
-    winrt::Windows::Foundation::IInspectable Current##name()                                                                                             \
-    {                                                                                                                                                    \
-        return winrt::box_value<winrt::Microsoft::Terminal::Settings::Editor::EnumEntry>(_##name##Map.Lookup(settingsModelName().settingNameInModel())); \
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/Utils.h:63]: The only change here is that I removed the `()` from `settingsModelName`. Needed this because I didn't want to expose a getter to `_globals` in `GlobalSettingsViewModel`.

File: src/cascadia/TerminalSettingsEditor/ViewModelHelpers.h [hunk 9]
```diff
@@ -39,24 +39,28 @@ struct ViewModelHelper
...
-    template<typename T>                                  \
-    void name(const T& value)                             \
-    {                                                     \
-        if (name() != value)                              \
-        {                                                 \
-            target.name(value);                           \
-            _NotifyChanges(L"Has" #name, L## #name);      \
-        }                                                 \
-    }                                                     \
+#define _GET_SETTING_FUNC(target, name) \
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/ViewModelHelpers.h:42]: A bit annoyed by this change. I wanted to use `OBSERVABLE_PROJECTED_SETTING` for Default Profile (I think?), but the `ClearXXX` function wasn't valid. So I had to break things up a bit here.

File: src/cascadia/TerminalSettingsEditor/GlobalSettingsViewModel.h (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/GlobalSettingsViewModel.h:9]: I feel like it's fine. I think the traditional OS code style would be 
```
TerminalSettingsModel/ -> contains the actual soure
  - inc/ -> has shared headers like MTSMSettings.h
  - lib/ -> builds the .lib
  - dll/ -> builds the .dll
```

but honestly, _meh_

File: src/cascadia/TerminalSettingsEditor/ViewModelHelpers.h (hunk 9 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/ViewModelHelpers.h:40]: wait isn't that why we have the `PERMANENT` variant of `OBSERVABLE...`? Like, Name and Guid cannot be cleared.

File: src/cascadia/TerminalSettingsEditor/Launch.xaml (hunk 6 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/Launch.xaml:78]: meh, this is fine, I don't think it's much worse than it was before

File: src/cascadia/TerminalSettingsEditor/ViewModelHelpers.h (hunk 9 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/ViewModelHelpers.h:40]: yea this is probably the only one I'm not signing off over. Not blocking, over, but it is a little weird

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/ViewModelHelpers.h:40]: Sorry, the exact problem is that `DefaultProfile` only has a getter and a setter (because we do that dance where we get the unparsed default profile, but we _always_ have a default profile). So it doesn't use inheritance at all.

The `PERMANENT` variant gets rid of `ClearXXX`, but still uses the `HasXXX` API. I could introduce a third macro called `GET_SET_XXX` that would just expose the getter/setter. Idk how valuable that would be though tbh.
